ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -1674,172 +1672,188 @@ extension SignalProducerProtocol {
+	let capacity: Int
+
+	/// All cached values.
 	var values: [Value] = []
 
-	/// Any terminating event sent to the buffer.
+	/// A termination event emitted by the underlying producer.
 	///
 	/// This will be nil if termination has not occurred.
 	var terminationEvent: Event<Value, Error>?
 
-	/// The observers currently attached to the buffered producer, or nil if the
-	/// producer was terminated.
+	/// The observers currently attached to the caching producer, or `nil` if the
+	/// caching producer was terminated.
 	var observers: Bag<Signal<Value, Error>.Observer>? = Bag()
 
-	/// The set of unused replay token identifiers.
-	var replayBuffers: Bag<ReplayBuffer<Value>> = Bag()
+	/// The set of in-flight replay tokens.
+	var replayBuffers: [ObjectIdentifier: ReplayBuffer<Value>] = [:]
 
-	/// Appends a new value to the buffer, trimming it down to the given capacity
-	/// if necessary.
-	mutating func add(_ value: Value, upTo capacity: Int) {
-		precondition(capacity >= 0)
+	/// Initialize the replay state.
+	///
+	/// - parameters:
+	///   - capacity: The maximum amount of values which can be cached by the
+	///               replay state.
+	init(upTo capacity: Int) {
+		self.capacity = capacity
+	}
 
-		for buffer in replayBuffers {
-			buffer.values.append(value)
+	/// Attempt to observe the replay state.
+	///
+	/// - warning: Repeatedly observing the replay state with the same observer
+	///            should be avoided.
+	///
+	/// - parameters:
+	///   - observer: The observer to be registered.
+	///
+	/// - throws:
+	///   `ReplayError.pending(values:)` if there are still one or more values
+	///   pending for being replayed by the observer.
+	///
+	/// - returns:
+	///   The removal token of the observer, or `nil` if the replay state has
+	///   already terminated.
+	mutating func observe(_ observer: Signal<Value, Error>.Observer) throws -> RemovalToken? {
+		// The only use case is in `replayLazily` which has always a unique observer
+		// created for each produced signal. So we can use the ObjectIdentifier to
+		// track them directly.
+		let id = ObjectIdentifier(observer)
+
+		if let buffer = replayBuffers[id] {
+			if buffer.values.isEmpty {
+				replayBuffers.removeValue(forKey: id)
+				_ = terminationEvent.map(observer.action)
+				return observers?.insert(observer)
+			} else {
+				defer { buffer.values.removeAll() }
+				throw ReplayError<Value>.pending(values: buffer.values)
+			}
+		} else {
+			if values.isEmpty {
+				_ = terminationEvent.map(observer.action)
+				return observers?.insert(observer)
+			} else {
+				replayBuffers[id] = ReplayBuffer<Value>()
+				throw ReplayError<Value>.pending(values: values)
+			}
 		}
+	}
 
-		if capacity == 0 {
-			values = []
-			return
-		}
+	/// Enqueue the supplied event to the replay state.
+	///
+	/// - parameter:
+	///   - event: The event to be cached.
+	mutating func enqueue(_ event: Event<Value, Error>) {
+		if let value = event.value {
```

Comment:
I think it'd be preferable to use a `switch` here and explicitly list the cases in the `else` condition.